extension/persistence/relational-jdbc/src/main/java/org/apache/polaris/extension/persistence/relational/jdbc/JdbcBasePersistenceImpl.java

```diff
@@ -701,6 +712,180 @@ public void deletePrincipalSecrets(
     }
   }
 
+  @Override
+  public void writeToPolicyMappingRecords(
+      @Nonnull PolarisCallContext callCtx, @Nonnull PolarisPolicyMappingRecord record) {
+    try {
+      datasourceOperations.runWithinTransaction(
+          statement -> {
+            PolicyType policyType = PolicyType.fromCode(record.getPolicyTypeCode());
+            Preconditions.checkArgument(
+                policyType != null, "Invalid policy type code: %s", record.getPolicyTypeCode());
+
+            String query =
+                generateInsertQuery(
+                    ModelPolicyMappingRecord.fromPolicyMappingRecord(record), realmId);
+            if (policyType.isInheritable()) {
+              List<PolarisPolicyMappingRecord> existingRecords =
+                  loadPoliciesOnTargetByType(
+                      callCtx,
+                      record.getTargetCatalogId(),
+                      record.getTargetId(),
+                      record.getPolicyTypeCode());
+              if (existingRecords.size() > 1) {
+                throw new PolicyMappingAlreadyExistsException(existingRecords.getFirst());
+              } else if (existingRecords.size() == 1) {
+                PolarisPolicyMappingRecord existingRecord = existingRecords.getFirst();
+                if (existingRecord.getPolicyCatalogId() != record.getPolicyCatalogId()
+                    || existingRecord.getPolicyId() != record.getPolicyId()) {
+                  throw new PolicyMappingAlreadyExistsException(existingRecord);
+                }
```

Comment:
I think the logic would be like this:
```
if (inheritable) {
   // check if the mapping exist
  // throw if it does
} 

// insert or update a new mapping record.
```

Reply:
yup the same is happening, i structed one additional else as if its inheritable its gonna be an update.